ambry-frontend/src/main/java/com.github.ambry.frontend/AmbryBlobStorageService.java

```
@@ -606,45 +605,52 @@ public void onCompletion(final BlobInfo routerResult, Exception routerException)
+                  try {
+                    if (securityException == null) {
+                      if (subResource != null) {
+                        BlobInfo blobInfo = routerResult.getBlobInfo();
+                        Map<String, String> userMetadata = RestUtils.buildUserMetadata(blobInfo.getUserMetadata());
+                        if (userMetadata == null) {
+                          restResponseChannel.setHeader(RestUtils.Headers.CONTENT_TYPE, "application/octet-stream");
+                          restResponseChannel
+                              .setHeader(RestUtils.Headers.CONTENT_LENGTH, blobInfo.getUserMetadata().length);
+                          response = new ByteBufferReadableStreamChannel(ByteBuffer.wrap(blobInfo.getUserMetadata()));
+                        } else {
+                          setUserMetadataHeaders(userMetadata, restResponseChannel);
+                          restResponseChannel.setHeader(RestUtils.Headers.CONTENT_LENGTH, 0);
+                          response = new ByteBufferReadableStreamChannel(AmbryBlobStorageService.EMPTY_BUFFER);
+                        }
+                      } else if (!blobNotModified) {
+                        response = routerResult.getBlobDataChannel();
+                      }
+                    }
+                  } catch (Exception e) {
+                    frontendMetrics.getSecurityResponseCallbackProcessingError.inc();
+                    securityException = e;
+                  } finally {
+                    if (response != null || securityException != null || blobNotModified) {
+                      submitResponse(restRequest, restResponseChannel, response, securityException);
+                    }
+                    if (securityException != null && routerResult.getBlobDataChannel() != null) {
+                      try {
+                        routerResult.getBlobDataChannel().close();
+                      } catch (IOException ignored){
+                      }
```

A good rule of thumb I follow is that *only* the code that swallows exception *must* log it. This makes sure that 1)  exceptions aren't logged more than once 2) all exceptions are surely logged.

Also add a metric here.